File: src/vt_client.py
# src/vt_client.py
import json
import os
import socket
import subprocess
import threading
import time
from queue import Queue

class VTClient:
    def __init__(self, daemon_path="./bin/vt-daemon", socket_path="/tmp/vt.sock"):
        self.daemon_path = daemon_path
        self.socket_path = socket_path
        self.events = Queue()
        self._process = None
        self._socket = None
        self._stop_event = threading.Event()

    def start_daemon(self):
        if os.path.exists(self.socket_path):
            os.remove(self.socket_path)
        self._process = subprocess.Popen([self.daemon_path, "-s", self.socket_path])
        self._wait_for_socket()

    def connect(self):
        self._socket = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
        self._socket.connect(self.socket_path)
        self._stop_event.clear()
        threading.Thread(target=self._read_loop, daemon=True).start()

    def write(self, data):
        if not self._socket:
            raise RuntimeError("Not connected to daemon")
        payload = data.encode('utf-8') if isinstance(data, str) else data
        self._socket.sendall(payload)

    def health(self):
        try:
            with socket.socket(socket.AF_UNIX, socket.SOCK_STREAM) as sock:
                sock.connect(self.socket_path)
                sock.sendall(b"\x1B_VTD;health\x1B\\")
                sock.settimeout(1.0)
                with sock.makefile('r', encoding='utf-8') as f:
                    for line in f:
                        try:
                            parsed = json.loads(line)
                            if parsed.get("status") == "ok":
                                return True
                        except json.JSONDecodeError:
                            pass
        except Exception:
            pass
        return False

    def shutdown(self):
        try:
            with socket.socket(socket.AF_UNIX, socket.SOCK_STREAM) as sock:
                sock.connect(self.socket_path)
                sock.sendall(b"\x1B_VTD;shutdown\x1B\\")
        except Exception:
            pass
        if self._process:
            self._process.wait()

    def stop(self):
        self._stop_event.set()
        if self._socket:
            try:
                self._socket.shutdown(socket.SHUT_RDWR)
                self._socket.close()
            except Exception:
                pass
            self._socket = None

    def _read_loop(self):
        try:
            with self._socket.makefile('r', encoding='utf-8') as f:
                while not self._stop_event.is_set():
                    line = f.readline()
                    if not line:
                        break
                    try:
                        self.events.put(json.loads(line))
                    except json.JSONDecodeError:
                        pass
        except Exception:
            pass

    def _wait_for_socket(self):
        for _ in range(30):
            if os.path.exists(self.socket_path):
                return
            time.sleep(0.1)
        raise RuntimeError(f"Daemon failed to start or bind to socket at {self.socket_path}")

File: src/vt_client.rb
# src/vt_client.rb
require 'socket'
require 'json'
require 'thread'

class VTClient
  attr_reader :socket_path, :daemon_path, :events

  def initialize(daemon_path = './bin/vt-daemon', socket_path = '/tmp/vt.sock')
    @daemon_path = daemon_path
    @socket_path = socket_path
    @events = Queue.new
    @process = nil
    @socket = nil
    @read_thread = nil
  end

  def start_daemon
    File.delete(@socket_path) if File.exist?(@socket_path)
    @process = Process.spawn(@daemon_path, '-s', @socket_path)
    wait_for_socket
  end

  def connect
    @socket = UNIXSocket.new(@socket_path)
    @read_thread = Thread.new { read_loop }
  end

  def write(data)
    raise "Not connected to daemon" unless @socket
    @socket.write(data)
    @socket.flush
  end

  def health?
    with_temp_socket do |sock|
      sock.print("\x1B_VTD;health\x1B\\")
      sock.flush
      while (line = sock.gets)
        begin
          json = JSON.parse(line)
          return true if json["status"] == "ok"
        rescue JSON::ParserError
        end
      end
    end
    false
  rescue StandardError
    false
  end

  def shutdown
    with_temp_socket do |sock|
      sock.print("\x1B_VTD;shutdown\x1B\\")
      sock.flush
    end
    Process.wait(@process) if @process
  rescue StandardError
  end

  def stop
    @events.close unless @events.closed?
    @socket&.close
    @socket = nil
    @read_thread&.kill
  end

  private

  def read_loop
    return unless @socket
    while !@socket.closed? && (line = @socket.gets)
      begin
        @events.push(JSON.parse(line))
      rescue JSON::ParserError
      rescue ClosedQueueError
        break
      end
    end
  rescue IOError
  ensure
    @events.close unless @events.closed?
  end

  def with_temp_socket
    sock = UNIXSocket.new(@socket_path)
    begin
      yield sock
    ensure
      sock.close
    end
  end

  def wait_for_socket
    30.times do
      return if File.exist?(@socket_path)
      sleep 0.1
    end
    raise "Daemon failed to start or bind to socket at #{@socket_path}"
  end
end

